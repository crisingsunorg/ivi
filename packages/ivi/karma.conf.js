const webpack = require("webpack");

module.exports = function (config) {
  config.set({
    browsers: ["ChromeHeadless"],
    frameworks: ["mocha"],
    reporters: ["mocha"],
    preprocessors: { "tests/index.ts": ["webpack", "sourcemap"] },
    files: ["tests/index.ts"],

    colors: true,
    autoWatch: true,

    webpack: {
      module: {
        rules: [
          {
            test: /\.ts$/,
            exclude: /node_modules/,
            use: [
              {
                loader: "ts-loader",
                options: {
                  configFileName: "tsconfig.tests.json",
                },
              },
            ],
          },
        ],
      },
      plugins: [
        new webpack.DefinePlugin({
          "__IVI_DEV__": true,
          "__IVI_BROWSER__": true,
        }),
        new webpack.SourceMapDevToolPlugin({
          test: /\.(ts|js)$/,
        }),
      ],
      resolve: {
        extensions: [".ts", ".js"],
      },
      performance: {
        hints: false
      },
    },

    webpackMiddleware: {
      stats: "errors-only",
      noInfo: true
    },

    mochaReporter: {
      showDiff: true,
    },

    client: {
      mocha: {
        reporter: "html",
        ui: "bdd",
      }
    },

    mime: {
      "text/x-typescript": ["ts"],
    },
  });
};
