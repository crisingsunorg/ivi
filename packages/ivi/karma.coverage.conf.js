const path = require("path");
const webpack = require("webpack");
const merge = require("webpack-merge");
const baseConfig = require("./karma.conf");

module.exports = function (config) {
  baseConfig(config);

  config.set({
    singleRun: true,

    webpack: merge.smart(config.webpack, {
      module: {
        rules: [{
          test: /\.ts$/,
          include: path.resolve("src/"),
          exclude: [
            /node_modules/,
            path.resolve("src/dev_mode/dev_mode.ts"),
            path.resolve("src/dev_mode/stack_trace.ts"),
          ],
          loader: "istanbul-instrumenter-loader",
          enforce: "post",
        }],
      },
    }),

    reporters: ["coverage-istanbul"],

    coverageIstanbulReporter: {
      reports: ["lcov", "text"],
      dir: "./coverage",
      fixWebpackSourcePath: true,
      skipFilesWithNoCoverage: true,
    },
  });

  if (process.env.TRAVIS) {
    config.coverageIstanbulReporter.reports = ["lcov"];
  }
};
