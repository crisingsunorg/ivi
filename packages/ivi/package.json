{
  "name": "ivi",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "es2016": "dist/es2016/index.js",
  "unpkg": "dist/unpkg/ivi.js",
  "typings": "dist/typings/index.d.ts",
  "sideEffects": false,
  "version": "0.10.0",
  "description": "Javascript (TypeScript) library for building web user interfaces.",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ui",
    "virtualdom",
    "vdom",
    "diff",
    "browser",
    "typescript"
  ],
  "homepage": "https://github.com/ivijs/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/ivijs/ivi"
  },
  "bugs": {
    "url": "https://github.com/ivijs/ivi/issues"
  },
  "github": "https://github.com/ivijs/ivi",
  "files": [
    "dist",
    "src",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module && npm run dist:es2016 && npm run dist:unpkg",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json",
    "dist:es2016": "../../node_modules/.bin/tsc -p ./tsconfig.build.es2016.json",
    "dist:unpkg": "../../node_modules/.bin/webpack --config ./webpack.config.unpkg.js"
  },
  "dependencies": {},
  "peerDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  },
  "devDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  }
}
