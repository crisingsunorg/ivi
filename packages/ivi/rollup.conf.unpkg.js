import resolve from "rollup-plugin-node-resolve";

export default {
  name: "ivi",
  input: __dirname + "/unpkg/entry.js",
  output: {
    file: __dirname + "/dist/unpkg/index.js",
    format: "umd",
  },
  plugins: [
    resolve(),
  ],
};
