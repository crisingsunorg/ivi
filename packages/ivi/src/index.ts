/**
 * Common types and functions re-exported from ivi-core.
 */
export { KeyCode, KeyLocation, MouseButtons, SyncableValue, PROPERTY, UNSAFE_HTML, EVENT, _ } from "ivi-core";

// Virtual DOM
export { VNodeFlags } from "./vdom/flags";
export { StatelessComponent } from "./vdom/stateless_component";
export { StatefulComponent, Component, isComponentAttached } from "./vdom/stateful_component";
export { ConnectDescriptor } from "./vdom/connect_descriptor";
export {
  VNode,
  getDOMNode, getComponent,
  stopDirtyChecking,
} from "./vdom/vnode";
export { fragment, map, mapRange, mapIterable } from "./vdom/vnode_collections";
export { element } from "./vdom/element";
export { statefulComponent, statelessComponent, withShouldUpdate, context, connect } from "./vdom/vnode_factories";
export { invalidate, dirtyCheck, render } from "./vdom/scheduler";

// Events
export {
  EventHandlerFlags, SyntheticEventFlags, EventFlags, PREVENT_DEFAULT, STOP_PROPAGATION, NativeEventSourceFlags,
} from "./events/flags";
export {
  getEventTarget,
  EVENT_CAPTURE_PASSIVE_OPTIONS, EVENT_CAPTURE_ACTIVE_OPTIONS, EVENT_PASSIVE_OPTIONS, EVENT_ACTIVE_OPTIONS,
} from "./events/utils";
export { SyntheticEvent } from "./events/synthetic_event";
export { EventDispatcher } from "./events/event_dispatcher";
export { EventHandler } from "./events/event_handler";
export { DispatchTarget } from "./events/dispatch_target";
export { accumulateDispatchTargets } from "./events/accumulate_dispatch_targets";
export { dispatchEvent } from "./events/dispatch_event";
export { SyntheticNativeEvent } from "./events/synthetic_native_event";
export {
  NativeEventDispatcher, createNativeEventDispatcher,
  beforeNativeEvent, afterNativeEvent,
  removeBeforeNativeEvent, removeAfterNativeEvent,
} from "./events/native_event_dispatcher";
export {
  EVENT_DISPATCHER_ABORT, EVENT_DISPATCHER_ACTIVATE, EVENT_DISPATCHER_ARIA_REQUEST, EVENT_DISPATCHER_BEFORE_ACTIVATE,
  EVENT_DISPATCHER_BEFORE_COPY, EVENT_DISPATCHER_BEFORE_CUT, EVENT_DISPATCHER_BEFORE_DEACTIVATE,
  EVENT_DISPATCHER_BEFORE_PASTE, EVENT_DISPATCHER_BLUR, EVENT_DISPATCHER_CAN_PLAY, EVENT_DISPATCHER_CAN_PLAYTHROUGH,
  EVENT_DISPATCHER_CHANGE, EVENT_DISPATCHER_CLICK, EVENT_DISPATCHER_CONTEXT_MENU, EVENT_DISPATCHER_COPY,
  EVENT_DISPATCHER_CUE_CHANGE, EVENT_DISPATCHER_CUT, EVENT_DISPATCHER_DOUBLE_CLICK, EVENT_DISPATCHER_DEACTIVATE,
  EVENT_DISPATCHER_DRAG, EVENT_DISPATCHER_DRAG_END, EVENT_DISPATCHER_DRAG_ENTER, EVENT_DISPATCHER_DRAG_LEAVE,
  EVENT_DISPATCHER_DRAG_OVER, EVENT_DISPATCHER_DRAG_START, EVENT_DISPATCHER_DROP, EVENT_DISPATCHER_DURATION_CHANGE,
  EVENT_DISPATCHER_EMPTIED, EVENT_DISPATCHER_ENCRYPTED, EVENT_DISPATCHER_ENDED, EVENT_DISPATCHER_ERROR,
  EVENT_DISPATCHER_FOCUS, EVENT_DISPATCHER_GOT_POINTER_CAPTURE, EVENT_DISPATCHER_INPUT, EVENT_DISPATCHER_INVALID,
  EVENT_DISPATCHER_KEY_DOWN, EVENT_DISPATCHER_KEY_PRESS, EVENT_DISPATCHER_KEY_UP, EVENT_DISPATCHER_LOAD,
  EVENT_DISPATCHER_LOADED_DATA, EVENT_DISPATCHER_LOADED_METADATA, EVENT_DISPATCHER_LOAD_START,
  EVENT_DISPATCHER_LOST_POINTER_CAPTURE, EVENT_DISPATCHER_MOUSE_DOWN, EVENT_DISPATCHER_MOUSE_ENTER,
  EVENT_DISPATCHER_MOUSE_LEAVE, EVENT_DISPATCHER_MOUSE_MOVE, EVENT_DISPATCHER_MOUSE_OUT, EVENT_DISPATCHER_MOUSE_OVER,
  EVENT_DISPATCHER_MOUSE_UP, EVENT_DISPATCHER_PASTE, EVENT_DISPATCHER_PAUSE, EVENT_DISPATCHER_PLAY,
  EVENT_DISPATCHER_PLAYING, EVENT_DISPATCHER_POINTER_CANCEL, EVENT_DISPATCHER_POINTER_DOWN,
  EVENT_DISPATCHER_POINTER_ENTER, EVENT_DISPATCHER_POINTER_LEAVE, EVENT_DISPATCHER_POINTER_MOVE,
  EVENT_DISPATCHER_POINTER_OUT, EVENT_DISPATCHER_POINTER_OVER, EVENT_DISPATCHER_POINTER_UP, EVENT_DISPATCHER_PROGRESS,
  EVENT_DISPATCHER_RATE_CHANGE, EVENT_DISPATCHER_RESET, EVENT_DISPATCHER_SCROLL, EVENT_DISPATCHER_SEEKED,
  EVENT_DISPATCHER_SEEKING, EVENT_DISPATCHER_SELECT, EVENT_DISPATCHER_SELECT_START, EVENT_DISPATCHER_STALLED,
  EVENT_DISPATCHER_SUBMIT, EVENT_DISPATCHER_SUSPEND, EVENT_DISPATCHER_TIME_UPDATE, EVENT_DISPATCHER_TOUCH_CANCEL,
  EVENT_DISPATCHER_TOUCH_END, EVENT_DISPATCHER_TOUCH_MOVE, EVENT_DISPATCHER_TOUCH_START, EVENT_DISPATCHER_UNLOAD,
  EVENT_DISPATCHER_VOLUME_CHANGE, EVENT_DISPATCHER_WAITING, EVENT_DISPATCHER_WHEEL, EVENT_DISPATCHER_ACTIVE_TOUCH_END,
  EVENT_DISPATCHER_ACTIVE_TOUCH_MOVE, EVENT_DISPATCHER_ACTIVE_TOUCH_START, EVENT_DISPATCHER_ACTIVE_WHEEL,

  createNativeEventHandler,

  onAbort, onActivate, onAriaRequest, onBeforeActivate, onBeforeCopy, onBeforeCut, onBeforeDeactivate, onBeforePaste,
  onBlur, onCanPlay, onCanPlaythrough, onChange, onClick, onContextMenu, onCopy, onCueChange, onCut, onDoubleClick,
  onDeactivate, onDrag, onDragEnd, onDragEnter, onDragLeave, onDragOver, onDragStart, onDrop, onDurationChange,
  onEmptied, onEncrypted, onEnded, onError, onFocus, onGotPointerCapture, onInput, onInvalid, onKeyDown, onKeyPress,
  onKeyUp, onLoad, onLoadedData, onLoadedMetadata, onLoadStart, onLostPointerCapture, onMouseDown, onMouseEnter,
  onMouseLeave, onMouseMove, onMouseOut, onMouseOver, onMouseUp, onPaste, onPause, onPlay, onPlaying, onPointerCancel,
  onPointerDown, onPointerEnter, onPointerLeave, onPointerMove, onPointerOut, onPointerOver, onPointerUp, onProgress,
  onRateChange, onReset, onScroll, onSeeked, onSeeking, onSelect, onSelectStart, onStalled, onSubmit, onSuspend,
  onTimeUpdate, onTouchCancel, onTouchEnd, onTouchMove, onTouchStart, onUnload, onVolumeChange, onWaiting, onWheel,
  onActiveTouchEnd, onActiveTouchMove, onActiveTouchStart, onActiveWheel,
} from "./events/native_events";
