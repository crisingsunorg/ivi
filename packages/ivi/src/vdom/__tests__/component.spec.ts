import { useResetDOM, useResetModules, useDOMElement, useIVI, useTest, useComputedValue, useHTML } from "ivi-jest";
import { Op } from "ivi";
import { useDOMOpsCounters, useLifecycleCounters } from "./jest";

useResetDOM();
useResetModules();
const root = useDOMElement();
const domOps = useDOMOpsCounters();
const ivi = useIVI();
const h = useHTML();
const t = useTest();
const r = (op: Op) => t.render(op, root()).domNode;

describe("component", () => {
  describe("stateless", () => {
    const Component = useComputedValue(() => ivi.statelessComponent((op: Op) => op));

    describe("mount", () => {
      test("null root node", () => {
        const n = r(Component(null));
        expect(n).toMatchSnapshot();
      });

      test("basic root node", () => {
        const n = r(Component(123));
        expect(n).toMatchSnapshot();
      });

      test("component root node", () => {
        const n = r(Component(Component(123)));
        expect(n).toMatchSnapshot();
      });

      test("fragment root node", () => {
        const n = r(Component([1, 2, 3]));
        expect(n).toMatchSnapshot();
      });
    });

    describe("update", () => {
      describe("root node", () => {
        test("1", () => {
          const a = r(h.div());
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("2", () => {
          const a = r(Component(h.div()));
          const b = r(h.div());
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("3", () => {
          const a = r(Component(h.div()));
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("4", () => {
          const a = r(Component(null));
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("5", () => {
          const a = r(Component(h.div()));
          const b = r(Component(null));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("6", () => {
          const n = h.div();
          const a = r(Component(n));
          const b = r(Component(n));
          expect(b).toMatchSnapshot();
          expect(a).toBe(b);
          expect(domOps()).toMatchSnapshot();
        });
      });

      describe("move", () => {
        test("1", () => {
          r(ivi.TrackByKey([
            ivi.key(0, Component(0)),
            ivi.key(1, Component(1)),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, Component(1)),
            ivi.key(0, Component(0)),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });

        test("2", () => {
          const c0 = Component(0);
          const c1 = Component(1);
          r(ivi.TrackByKey([
            ivi.key(0, c0),
            ivi.key(1, c1),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, c1),
            ivi.key(0, c0),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });

        test("3", () => {
          const c0 = Component(0);
          const c1 = Component(null);
          r(ivi.TrackByKey([
            ivi.key(0, c0),
            ivi.key(1, c1),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, c1),
            ivi.key(0, c0),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });
      });
    });
  });

  describe("stateful", () => {
    const Component = useComputedValue(() => ivi.component(() => (op: Op) => op));

    describe("mount", () => {
      test("null root node", () => {
        const n = r(Component(null));
        expect(n).toMatchSnapshot();
      });

      test("basic root node", () => {
        const n = r(Component(123));
        expect(n).toMatchSnapshot();
      });

      test("component root node", () => {
        const n = r(Component(Component(123)));
        expect(n).toMatchSnapshot();
      });

      test("fragment root node", () => {
        const n = r(Component([1, 2, 3]));
        expect(n).toMatchSnapshot();
      });
    });

    describe("update", () => {
      describe("root node", () => {
        test("1", () => {
          const a = r(h.div());
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("2", () => {
          const a = r(Component(h.div()));
          const b = r(h.div());
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("3", () => {
          const a = r(Component(h.div()));
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("4", () => {
          const a = r(Component(null));
          const b = r(Component(h.div()));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("5", () => {
          const a = r(Component(h.div()));
          const b = r(Component(null));
          expect(b).toMatchSnapshot();
          expect(a).not.toBe(b);
          expect(domOps()).toMatchSnapshot();
        });

        test("6", () => {
          const n = h.div();
          const a = r(Component(n));
          const b = r(Component(n));
          expect(b).toMatchSnapshot();
          expect(a).toBe(b);
          expect(domOps()).toMatchSnapshot();
        });
      });

      describe("move", () => {
        test("1", () => {
          r(ivi.TrackByKey([
            ivi.key(0, Component(0)),
            ivi.key(1, Component(1)),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, Component(1)),
            ivi.key(0, Component(0)),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });

        test("2", () => {
          const c0 = Component(0);
          const c1 = Component(1);
          r(ivi.TrackByKey([
            ivi.key(0, c0),
            ivi.key(1, c1),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, c1),
            ivi.key(0, c0),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });

        test("3", () => {
          const c0 = Component(0);
          const c1 = Component(null);
          r(ivi.TrackByKey([
            ivi.key(0, c0),
            ivi.key(1, c1),
          ]));
          r(ivi.TrackByKey([
            ivi.key(1, c1),
            ivi.key(0, c0),
          ]));
          expect(root()).toMatchSnapshot();
          expect(domOps()).toMatchSnapshot();
        });
      });
    });

    describe("hooks", () => {
      describe("shouldUpdate", () => {
        test("props", () => {
          const shouldUpdate = jest.fn(() => true);
          const C = ivi.component(() => () => null, shouldUpdate);

          r(C(1));
          r(C(2));

          expect(shouldUpdate).toBeCalledTimes(1);
          expect(shouldUpdate).toBeCalledWith(1, 2);
        });
      });

      describe("useUnmount", () => {
        test("one hook", () => {
          const unmount = jest.fn();
          const C = ivi.component((c) => {
            ivi.useUnmount(c, unmount);
            return () => null;
          });
          r(C());
          r(null);
          expect(unmount).toBeCalledTimes(1);
          expect(unmount).toBeCalledWith(true);
        });
      });

      test("two hooks", () => {
        const unmount1 = jest.fn();
        const unmount2 = jest.fn();
        const C = ivi.component((c) => {
          ivi.useUnmount(c, unmount1);
          ivi.useUnmount(c, unmount2);
          return () => null;
        });
        r(C());
        r(null);
        expect(unmount1).toBeCalledTimes(1);
        expect(unmount1).toBeCalledWith(true);
        expect(unmount2).toBeCalledTimes(1);
        expect(unmount2).toBeCalledWith(true);
      });
    });

    describe("lifecycle", () => {
      const Static = useComputedValue(() => ivi.component(() => (op: Op) => op, () => false));
      const lifecycle = useLifecycleCounters();
      const createLifecycleTester = (id: string) => (ivi.component<Op>(
        (component) => {
          lifecycle.touch(id, "constructor");
          const mutation = ivi.useMutationEffect(component, () => { lifecycle.touch(id, "mutationEffect"); });
          const layout = ivi.useLayoutEffect(component, () => { lifecycle.touch(id, "layoutEffect"); });
          const effect = ivi.useEffect(component, () => { lifecycle.touch(id, "effect"); });
          ivi.useUnmount(component, () => { lifecycle.touch(id, "unmount"); });
          return (child) => (mutation(), layout(), effect(), lifecycle.touch(id, "render"), child);
        },
        (prev, next) => (lifecycle.touch(id, "shouldUpdate"), true),
      ));

      test("1", () => {
        r(createLifecycleTester("1")(null));

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("1", "mutationEffect")).toBe(2);
        expect(lifecycle.get("1", "layoutEffect")).toBe(3);
        expect(lifecycle.get("1", "effect")).toBe(4);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);
        expect(lifecycle.get("1", "unmount")).toBe(-1);
      });

      test("2", () => {
        r(createLifecycleTester("1")(null));
        r(null);

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("1", "mutationEffect")).toBe(2);
        expect(lifecycle.get("1", "layoutEffect")).toBe(3);
        expect(lifecycle.get("1", "effect")).toBe(4);
        expect(lifecycle.get("1", "unmount")).toBe(5);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);
      });

      test("3", () => {
        r(createLifecycleTester("1")(createLifecycleTester("2")(null)));

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("2", "constructor")).toBe(2);
        expect(lifecycle.get("2", "render")).toBe(3);
        expect(lifecycle.get("1", "mutationEffect")).toBe(4);
        expect(lifecycle.get("2", "mutationEffect")).toBe(5);
        expect(lifecycle.get("1", "layoutEffect")).toBe(6);
        expect(lifecycle.get("2", "layoutEffect")).toBe(7);
        expect(lifecycle.get("1", "effect")).toBe(8);
        expect(lifecycle.get("2", "effect")).toBe(9);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);
        expect(lifecycle.get("1", "unmount")).toBe(-1);

        expect(lifecycle.get("2", "shouldUpdate")).toBe(-1);
        expect(lifecycle.get("2", "unmount")).toBe(-1);
      });

      test("4", () => {
        r(createLifecycleTester("1")(createLifecycleTester("2")(null)));
        r(null);

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("2", "constructor")).toBe(2);
        expect(lifecycle.get("2", "render")).toBe(3);
        expect(lifecycle.get("1", "mutationEffect")).toBe(4);
        expect(lifecycle.get("2", "mutationEffect")).toBe(5);
        expect(lifecycle.get("1", "layoutEffect")).toBe(6);
        expect(lifecycle.get("2", "layoutEffect")).toBe(7);
        expect(lifecycle.get("1", "effect")).toBe(8);
        expect(lifecycle.get("2", "effect")).toBe(9);
        expect(lifecycle.get("2", "unmount")).toBe(10);
        expect(lifecycle.get("1", "unmount")).toBe(11);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);

        expect(lifecycle.get("2", "shouldUpdate")).toBe(-1);
      });

      test("5", () => {
        const lc = createLifecycleTester("1")(null);
        r(lc);
        r(lc);

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("1", "mutationEffect")).toBe(2);
        expect(lifecycle.get("1", "layoutEffect")).toBe(3);
        expect(lifecycle.get("1", "effect")).toBe(4);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);
        expect(lifecycle.get("1", "unmount")).toBe(-1);
      });

      test("6", () => {
        r(Static(createLifecycleTester("1")(null)));
        r(Static(createLifecycleTester("1")(null)));

        expect(lifecycle.get("1", "constructor")).toBe(0);
        expect(lifecycle.get("1", "render")).toBe(1);
        expect(lifecycle.get("1", "mutationEffect")).toBe(2);
        expect(lifecycle.get("1", "layoutEffect")).toBe(3);
        expect(lifecycle.get("1", "effect")).toBe(4);

        expect(lifecycle.get("1", "shouldUpdate")).toBe(-1);

        expect(lifecycle.get("1", "unmount")).toBe(-1);
      });
    });
  });
});
