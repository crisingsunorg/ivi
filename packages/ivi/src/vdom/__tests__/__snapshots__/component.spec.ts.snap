// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component stateful mount basic root node 1`] = `123`;

exports[`component stateful mount component root node 1`] = `123`;

exports[`component stateful mount fragment root node 1`] = `1`;

exports[`component stateful mount null root node 1`] = `null`;

exports[`component stateful update move 1 1`] = `
<div>
  1
  0
</div>
`;

exports[`component stateful update move 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 2,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update move 2 1`] = `
<div>
  1
  0
</div>
`;

exports[`component stateful update move 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 2,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update move 3 1`] = `
<div>
  0
</div>
`;

exports[`component stateful update move 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 1 1`] = `<div />`;

exports[`component stateful update root node 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 2 1`] = `<div />`;

exports[`component stateful update root node 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 3 1`] = `<div />`;

exports[`component stateful update root node 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 4 1`] = `<div />`;

exports[`component stateful update root node 4 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 5 1`] = `null`;

exports[`component stateful update root node 5 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful update root node 6 1`] = `<div />`;

exports[`component stateful update root node 6 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateful useSelect return value 1`] = `1`;

exports[`component stateless mount basic root node 1`] = `123`;

exports[`component stateless mount component root node 1`] = `123`;

exports[`component stateless mount fragment root node 1`] = `1`;

exports[`component stateless mount null root node 1`] = `null`;

exports[`component stateless update move 1 1`] = `
<div>
  1
  0
</div>
`;

exports[`component stateless update move 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 2,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update move 2 1`] = `
<div>
  1
  0
</div>
`;

exports[`component stateless update move 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 2,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update move 3 1`] = `
<div>
  0
</div>
`;

exports[`component stateless update move 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 1 1`] = `<div />`;

exports[`component stateless update root node 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 2 1`] = `<div />`;

exports[`component stateless update root node 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 3 1`] = `<div />`;

exports[`component stateless update root node 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 4 1`] = `<div />`;

exports[`component stateless update root node 4 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 5 1`] = `null`;

exports[`component stateless update root node 5 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`component stateless update root node 6 1`] = `<div />`;

exports[`component stateless update root node 6 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;
